### PR TITLE
feat(service): symbol recovery endpoints

### PR DESCRIPTION
**Pull request by LefKok**

### CODE REVIEW

File: crates/walrus-core/src/test_utils.rs [hunk 1]
```diff
@@ -31,11 +31,16 @@ pub fn keypair() -> BLS12381KeyPair {
     BLS12381KeyPair::generate(&mut rng)
 }
 
-/// Returns an arbitrary sliver for testing.
+/// Returns an arbitrary primary sliver for testing.
 pub fn sliver() -> Sliver {
```
Comment (asonnino) [inline crates/walrus-core/src/test_utils.rs]:
```suggestion
pub fn primary_sliver() -> Sliver {
```

File: crates/walrus-core/src/test_utils.rs [hunk 2]
```diff
@@ -31,11 +31,16 @@ pub fn keypair() -> BLS12381KeyPair {
...
     Sliver::Primary(encoding::Sliver::new([1, 2, 3, 4], 2, 1))
 }
 
+/// Returns an arbitrary secondary sliver for testing.
+pub fn sliver_secondary() -> Sliver {
```
Comment (asonnino) [inline crates/walrus-core/src/test_utils.rs]:
```suggestion
pub fn secondary_sliver() -> Sliver {
```

File: crates/walrus-service/src/node.rs [hunk 3]
```diff
@@ -39,6 +40,20 @@ pub enum RetrieveSliverError {
     Internal(#[from] anyhow::Error),
 }
 
+#[derive(Debug, thiserror::Error)]
+pub enum RetrieveSymbolError {
+    #[error("Invalid shard {0:?}")]
+    InvalidShard(ShardIndex),
+    #[error("Symbol Recovery Failed")]
+    RecoveryError,
```
Comment (asonnino) [inline crates/walrus-service/src/node.rs]: Do we have an easy way to pass the index or symbol reference to this error?

File: crates/walrus-service/src/node.rs [hunk 4]
```diff
@@ -39,6 +40,20 @@ pub enum RetrieveSliverError {
...
+    #[error("Sliver is not Available for Recovery")]
+    UnavailableSliver,
```
Comment (asonnino) [inline crates/walrus-service/src/node.rs]: Do we have an easy way to add the sliver index or reference to this error?

File: crates/walrus-service/src/node.rs [hunk 5]
```diff
@@ -39,6 +40,20 @@ pub enum RetrieveSliverError {
...
+    #[error("Symbol Recovering the Wrong Axis")]
+    WrongAxis,
```
Comment (asonnino) [inline crates/walrus-service/src/node.rs]: Do we have an easy way to specify the axis we used?

File: crates/walrus-service/src/server.rs [hunk 6]
```diff
@@ -235,6 +251,100 @@ impl<S: ServiceState + Send + Sync + 'static> UserServer<S> {
...
+    ) -> (
+        StatusCode,
+        Json<ServiceResponse<DecodingSymbol<walrus_core::encoding::Primary, MerkleProof>>>,
+    ) {
+        let Ok(blob_id) = BlobId::from_str(&encoded_blob_id) else {
+            tracing::debug!("Invalid blob ID {encoded_blob_id}");
+            return ServiceResponse::serialized_error(StatusCode::BAD_REQUEST, "Invalid blob ID");
+        };
+
+        match state.retrieve_sliver(&blob_id, sliver_pair_idx, SliverType::Secondary) {
```
Comment (asonnino) [inline crates/walrus-service/src/server.rs]: We usually try to do most of the heavy lifting inside node.rs and keep server.rs as simple as possible (we never know, we may move out of HTTP one day) but its is a matter of taste.

File: crates/walrus-sui/src/utils.rs [hunk 7]
```diff
@@ -178,19 +178,19 @@ pub(crate) fn blob_id_from_u256(input: U256) -> BlobId {
 }
 
 macro_rules! call_arg_pure {
-    ($value: expr) => {
+    ($value:expr) => {
```
Comment (asonnino) [inline crates/walrus-sui/src/utils.rs:181]: I don't know why your fmt did that, but as long as the CI passes it should be good

Comment (mlegner) [reply crates/walrus-sui/src/utils.rs:181]: I believe the canonical formatting is the one without space. It seems that the Rust formatter doesn't want to modify macro definitions very much though.

File: crates/walrus-core/src/encoding/symbols.rs [hunk 8]
```diff
@@ -280,6 +280,12 @@ pub struct DecodingSymbol<T: EncodingAxis, U = ()> {
     _axis: PhantomData<T>,
 }
 
+/// A primary decoding symbol to recover a primary sliver
+pub type PrimaryDecSymbol = DecodingSymbol<Primary>;
+
+///  A secondary decoding symbol to recover a secondary sliver
+pub type SecondaryDecSymbol = DecodingSymbol<Secondary>;
+
```
Comment (mlegner) [inline crates/walrus-core/src/encoding/symbols.rs]: Nit: Can we call them `PrimaryDecodingSymbol` and `SecondaryDecodingSymbol`? I think this isn't too long and the abbreviation is not obvious.
```suggestion
/// A primary decoding symbol to recover a primary sliver.
pub type PrimaryDecodingSymbol = DecodingSymbol<Primary>;

/// A secondary decoding symbol to recover a secondary sliver.
pub type SecondaryDecodingSymbol = DecodingSymbol<Secondary>;

```

File: crates/walrus-core/src/lib.rs [hunk 9]
```diff
@@ -211,6 +217,49 @@ impl Sliver {
...
 }
 
+/// A decoding symbol for recovering a sliver
+///
+/// Can be either a [`PrimaryDecSymbol`] or [`SecondaryDecSymbol`].
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub enum DecodingSymbol {
+    /// A primary decoding symbol to recover a primary sliver
+    Primary(PrimaryDecSymbol),
+    ///    A secondary decoding symbol to recover a secondary sliver
```
Comment (mlegner) [inline crates/walrus-core/src/lib.rs]: Nit:
```suggestion
    /// A secondary decoding symbol to recover a secondary sliver.
```

File: crates/walrus-core/src/lib.rs [hunk 10]
```diff
@@ -221,6 +270,16 @@ pub enum SliverType {
     Secondary,
 }
 
+impl SliverType {
+    /// Returns the opposite sliver type.
+    pub fn reverse(&self) -> SliverType {
```
Comment (mlegner) [inline crates/walrus-core/src/lib.rs]: Remark: In the `EncodingAxis` trait, we call this "orthogonal" instead of "opposite" or "reverse".

File: crates/walrus-core/src/test_utils.rs [hunk 11]
```diff
@@ -35,6 +36,11 @@ pub fn keypair() -> BLS12381KeyPair {
 pub fn sliver() -> Sliver {
     Sliver::Primary(encoding::Sliver::new([1, 2, 3, 4], 2, 1))
 }
+/// Returns an arbitrary decoding symbol for testing.
+
```
Comment (mlegner) [inline crates/walrus-core/src/test_utils.rs]: Nit:
```suggestion

/// Returns an arbitrary decoding symbol for testing.
```

File: crates/walrus-service/src/node.rs [hunk 12]
```diff
@@ -94,6 +107,15 @@ pub trait ServiceState {
...
     ) -> impl Future<Output = Result<Option<StorageConfirmation>, anyhow::Error>> + Send;
+
+    /// Retrieves a recovery symbol for a shard held by this storage node.
+    fn retrieve_recovery_symbol(
+        &self,
+        blob_id: &BlobId,
+        sliver_pair_idx: u16,
+        sliver_type: SliverType,
+        index: u32,
+    ) -> Result<Option<DecodingSymbol>, RetrieveSymbolError>;
```
Comment (mlegner) [inline crates/walrus-service/src/node.rs]: Suggestion: Indicate in the docstring or the variable names whether the `sliver_*` parameters are about the *target* (i.e., the sliver to be recovered) or the *source*. AFAIU, the `sliver_pair_index` is about the source while the `sliver_type` is about the target.

File: crates/walrus-service/src/node.rs [hunk 13]
```diff
@@ -250,6 +272,47 @@ impl ServiceState for StorageNode {
             Ok(None)
         }
     }
+
+    //TODO Add Proof
```
Comment (mlegner) [inline crates/walrus-service/src/node.rs]: Remark: Remaining TODO.

File: crates/walrus-service/src/node.rs [hunk 14]
```diff
@@ -250,6 +272,47 @@ impl ServiceState for StorageNode {
...
+    ) -> Result<Option<DecodingSymbol>, RetrieveSymbolError> {
+        let optional_sliver = self
+            .retrieve_sliver(blob_id, sliver_pair_idx, sliver_type.reverse())
+            .map_err(|_| {
+                RetrieveSymbolError::InvalidShard(shard_index_for_pair(
+                    sliver_pair_idx,
+                    self.n_shards.get(),
+                    blob_id,
+                ))
+            })?;
```
Comment (mlegner) [inline crates/walrus-service/src/node.rs]: Remark: Only the `RetrieveSliverError::InvalidShard` should be mapped to a `RetrieveSymbolError::InvalidShard`. 

Suggestion: Add
```rust
impl From<RetrieveSliverError> for RetrieveSymbolError {
    fn from(value: RetrieveSliverError) -> Self {
        match value {
            RetrieveSliverError::InvalidShard(s) => Self::InvalidShard(s),
            RetrieveSliverError::Internal(e) => Self::Internal(e),
        }
    }
}
```

then this can be simplified to
```suggestion
            .retrieve_sliver(blob_id, sliver_pair_idx, sliver_type.reverse())?;
```

File: crates/walrus-service/src/node.rs [hunk 15]
```diff
@@ -250,6 +272,47 @@ impl ServiceState for StorageNode {
...
+                    })?;
+                    let decoded_symbol = DecodingSymbol::Primary(symbol);
+                    Ok(Some(decoded_symbol))
+                }
+            },
+            None => Err(RetrieveSymbolError::UnavailableSliver(
+                sliver_pair_idx,
+                *blob_id,
+            )),
+        }
```
Comment (mlegner) [inline crates/walrus-service/src/node.rs]: How about this version to reduce the nesting?
```suggestion
        let Some(sliver) = optional_sliver else {
            return Err(RetrieveSymbolError::UnavailableSliver(
                sliver_pair_idx,
                *blob_id,
            ));
        };

        Ok(Some(match sliver {
            Sliver::Primary(inner) => {
                let symbol = inner.recovery_symbol_for_sliver(index).map_err(|_| {
                    RetrieveSymbolError::RecoveryError(index, sliver_pair_idx, *blob_id)
                })?;
                DecodingSymbol::Secondary(symbol)
            }
            Sliver::Secondary(inner) => {
                let symbol = inner.recovery_symbol_for_sliver(index).map_err(|_| {
                    RetrieveSymbolError::RecoveryError(index, sliver_pair_idx, *blob_id)
                })?;
                DecodingSymbol::Primary(symbol)
            }
        }))
```

File: crates/walrus-service/src/node.rs [hunk 16]
```diff
@@ -250,6 +272,47 @@ impl ServiceState for StorageNode {
...
     }
+
+    //TODO Add Proof
+    fn retrieve_recovery_symbol(
+        &self,
+        blob_id: &BlobId,
+        sliver_pair_idx: u16,
+        sliver_type: SliverType,
+        index: u32,
+    ) -> Result<Option<DecodingSymbol>, RetrieveSymbolError> {
```
Comment (mlegner) [inline crates/walrus-service/src/node.rs]: Remark: It seems to me that the `Option` here is not necessary.

Suggestion: Change the return type to `Result<DecodingSymbol, RetrieveSymbolError>`.

File: crates/walrus-service/src/server.rs [hunk 17]
```diff
@@ -202,6 +206,34 @@ impl<S: ServiceState + Send + Sync + 'static> UserServer<S> {
...
     }
 
+    async fn retrieve_recovery_symbol(
+        State(state): State<Arc<S>>,
+        Path((HexBlobId(blob_id), sliver_pair_idx, sliver_type, index)): Path<(
+            HexBlobId,
+            u16,
+            SliverType,
+            u32,
+        )>,
```
Comment (mlegner) [inline crates/walrus-service/src/server.rs:220]: Same remark about the source/target parameters as in `node.rs`.

File: crates/walrus-service/src/server.rs [hunk 18]
```diff
@@ -202,6 +206,34 @@ impl<S: ServiceState + Send + Sync + 'static> UserServer<S> {
...
+    ) -> (StatusCode, Json<ServiceResponse<Option<DecodingSymbol>>>) {
+        match state.retrieve_recovery_symbol(&blob_id, sliver_pair_idx, sliver_type, index) {
+            Ok(Some(symbol)) => {
+                tracing::debug!("Retrieved recovery symbol for {blob_id:?}");
+                ServiceResponse::serialized_success(StatusCode::OK, Some(symbol))
+            }
+            Ok(None) => {
+                tracing::debug!("Recovery symbol not found for {blob_id:?}");
+                ServiceResponse::serialized_not_found()
+            }
```
Comment (mlegner) [inline crates/walrus-service/src/server.rs]: Remark: This case is currently not applicable. OTOH, a `RetrieveSymbolError::InvalidShard` and/or `RetrieveSymbolError::UnavailableSliver` should probably result in a 404 instead of a 500.

File: crates/walrus-service/src/server.rs [hunk 19]
```diff
@@ -272,13 +304,27 @@ mod test {
         messages::StorageConfirmation,
         metadata::{UnverifiedBlobMetadataWithId, VerifiedBlobMetadataWithId},
         BlobId,
+        DecodingSymbol,
         Sliver,
         SliverType,
     };
 
-    use super::*;
```
Comment (mlegner) [inline crates/walrus-service/src/server.rs:279]: Question: Any reason not to keep the simple `use super::*`?

File: crates/walrus-service/src/server.rs [hunk 20]
```diff
@@ -615,4 +669,39 @@ mod test {
         cancel_token.cancel();
         handle.await.unwrap().unwrap();
     }
+
+    #[tokio::test]
+    async fn get_decoding_symbold() {
```
Comment (mlegner) [inline crates/walrus-service/src/server.rs]: Could we add a test for the negative cases (e.g., 404) as well?

File: crates/walrus-service/src/server.rs (hunk 19 above)
Comment (LefKok) [reply crates/walrus-service/src/server.rs:279]: Just overeager cleaning up

File: crates/walrus-service/src/node.rs [hunk 21]
```diff
@@ -39,6 +40,18 @@ pub enum RetrieveSliverError {
     Internal(#[from] anyhow::Error),
 }
 
+#[derive(Debug, thiserror::Error)]
+pub enum RetrieveSymbolError {
+    #[error("Invalid shard {0:?}")]
+    InvalidShard(ShardIndex),
+    #[error("Symbol Recovery Failed for Sliver {0:?}, Index {0:?} in Blob {2:?}")]
```
Comment (asonnino) [inline crates/walrus-service/src/node.rs]:
```suggestion
    #[error("Symbol recovery failed for sliver {0:?}, index {0:?} in blob {2:?}")]
```

File: crates/walrus-service/src/node.rs [hunk 22]
```diff
@@ -39,6 +40,18 @@ pub enum RetrieveSliverError {
...
+    RecoveryError(u32, u16, BlobId),
```
Comment (asonnino) [inline crates/walrus-service/src/node.rs:48]: This tuple may seem strange but a PR will change the type of the SliverIndex so it should make it nicer.

File: crates/walrus-service/src/node.rs [hunk 23]
```diff
@@ -39,6 +40,18 @@ pub enum RetrieveSliverError {
...
+    #[error("Sliver {0:?} is not Available for Recovery in Blob {1:?}")]
```
Comment (asonnino) [inline crates/walrus-service/src/node.rs]:
```suggestion
    #[error("Sliver {0:?} unavailable for recovery in blob {1:?}")]
```

File: crates/walrus-service/src/node.rs (hunk 13 above)
Comment (asonnino) [reply crates/walrus-service/src/node.rs]: We seem to use the following format:
```
    //TODO(@LefKok): Add Proof, #xxx
```
And then open open an issue #xxx to track this todo

File: crates/walrus-service/src/server.rs [hunk 24]
```diff
@@ -202,6 +206,34 @@ impl<S: ServiceState + Send + Sync + 'static> UserServer<S> {
...
 
+    async fn retrieve_recovery_symbol(
+        State(state): State<Arc<S>>,
+        Path((HexBlobId(blob_id), sliver_pair_idx, sliver_type, index)): Path<(
+            HexBlobId,
+            u16,
+            SliverType,
+            u32,
+        )>,
+    ) -> (StatusCode, Json<ServiceResponse<Option<DecodingSymbol>>>) {
```
Comment (asonnino) [inline crates/walrus-service/src/server.rs]: Do we need the Option here?

File: crates/walrus-service/src/server.rs [hunk 25]
```diff
@@ -30,6 +31,8 @@ pub const METADATA_ENDPOINT: &str = "/v1/blobs/:blobId/metadata";
 pub const SLIVER_ENDPOINT: &str = "/v1/blobs/:blobId/slivers/:sliverPairIdx/:sliverType";
 /// The path to get storage confirmations.
 pub const STORAGE_CONFIRMATION_ENDPOINT: &str = "/v1/blobs/:blobId/confirmation";
+/// The path to get recovery symbols.
+pub const RECOVERY_ENDPOINT: &str = "/v1/blobs/:blobId/slivers/:sliverPairIdx/:Type/:index";
```
Comment (asonnino) [inline crates/walrus-service/src/server.rs]:
```suggestion
pub const RECOVERY_ENDPOINT: &str = "/v1/blobs/:blobId/slivers/:sliverPairIdx/:sliverType/:index";
```

Comment (asonnino) [reply crates/walrus-service/src/server.rs]: What would be a more expressive name instead of ":index"?

Comment (LefKok) [reply crates/walrus-service/src/server.rs]: PeerIndex?

File: crates/walrus-service/src/server.rs (hunk 18 above)
Comment (LefKok) [reply crates/walrus-service/src/server.rs]: I just adopted the errors above. I will leave it the same and we can handle them all together

File: crates/walrus-service/src/server.rs (hunk 25 above)
Comment (mlegner) [reply crates/walrus-service/src/server.rs]: `targetSliverIndex` or `targetSliverPairIndex` (depending on which of them we want)?

File: crates/walrus-service/src/node.rs [hunk 26]
```diff
@@ -94,6 +116,18 @@ pub trait ServiceState {
         &self,
         blob_id: &BlobId,
     ) -> impl Future<Output = Result<Option<StorageConfirmation>, anyhow::Error>> + Send;
+
+    /// Retrieves a recovery symbol for a shard held by this storage node.
+    /// The sliver_type is the target type of the sliver that will be recovered
+    /// The sliver_pair_idx is the index of the sliver pair that we want to access
+    /// Index is the requesters index in the established order of storage nodes
```
Comment (mlegner) [inline crates/walrus-service/src/node.rs]: Remark: I don't fully understand this statement. What is the "established order of storage nodes"? The `encoding::slivers::recovery_symbol_for_sliver` function takes the `target_pair_idx` as a parameter and gets the `index`. So maybe just say
```suggestion
    /// The `index` is the pair index of the sliver to be recovered.
```

File: crates/walrus-service/src/server.rs (hunk 18 above)
Comment (mlegner) [reply crates/walrus-service/src/server.rs]: I created a new issue for this: #178

File: crates/walrus-service/src/server.rs (hunk 19 above)
Comment (mlegner) [reply crates/walrus-service/src/server.rs:279]: Not really important, but we have the `use super::*` in all other test modules.

File: crates/walrus-service/src/server.rs [hunk 27]
```diff
@@ -615,4 +672,64 @@ mod test {
...
+    #[tokio::test]
+    async fn decoding_symbol_not_found() {
+        let server = UserServer::new(Arc::new(MockServiceState), CancellationToken::new());
+        let test_private_parameters = test_utils::storage_node_private_parameters();
+        let _handle = tokio::spawn(async move {
+            let network_address = test_private_parameters.network_address;
+            server.run(&network_address).await
+        });
+
+        tokio::task::yield_now().await;
```
Comment (mlegner) [inline crates/walrus-service/src/server.rs:722]: Nit: This could be factored out into a separate function as it is used in multiple tests. Can be done in a separate PR to streamline tests more generally.

File: crates/walrus-service/src/server.rs [hunk 28]
```diff
@@ -615,4 +672,64 @@ mod test {
...
+
+        let sliver_pair_id = 1; // Triggers a not found response
+        let path = RECOVERY_ENDPOINT
+            .replace(":blobId", &blob_id.to_string())
+            .replace(":sliverPairIdx", &sliver_pair_id.to_string())
+            .replace(":sliverType", "primary")
+            .replace(":index", "0");
+        let url = format!("http://{}{path}", test_private_parameters.network_address);
+        let client = reqwest::Client::new();
+        let res = client.get(url).json(&blob_id).send().await.unwrap();
```
Comment (mlegner) [inline crates/walrus-service/src/server.rs:734]: Remark: Also a candidate for deduplication in a follow-up.

File: crates/walrus-service/src/server.rs (hunk 19 above)
Comment (LefKok) [reply crates/walrus-service/src/server.rs:279]: For some reason the linter complains and doesn't allow me to merge if I keep it

Comment (mlegner) [reply crates/walrus-service/src/server.rs:279]: OK, weird. Let's leave it like this, then I can take a look separately. 🙂
